libsolidity/codegen/CompilerUtils.h

```diff
@@ -65,6 +65,13 @@ class CompilerUtils
 	/// Stack post:
 	void revertWithStringData(Type const& _argumentType);
 
+	/// Computes the absolute calldata offset of a tail given a base reference and the (absolute)
+	/// offset of the tail pointer. Performs bounds checks. For dynamically sized types it also
```

Comment:
```suggestion
	/// offset of the tail pointer. Performs bounds checks. If @a _type is a dynamically sized array it also
```

Reply:
Actually I'm wondering whether this second part should not be in its own utility function - we might even already have it somewhere.

Reply:
Ok, I think the root problem here lies in the fact that we store the array length on the stack for calldata arrays. Once we change that, this might also get simpler.

Reply:
*Should* we change it? I'm still wondering whether we should allow referencing subranges of calldata arrays - that'd allow to pass them around more flexibly to internal functions at some point and it would solve the issue of ABI decoding ``msg.data``... and that's trivial with the length on the stack, but we probably can't do it otherwise.

Reply:
passing around subranges of calldata arrays does not sound very type-safe to me... I think `msg.data` should just be a different type - and that type might then also be used for subranges of calldata arrays.

Reply:
Ah - so a "full array" type would only store a single reference on the stack that points to the length, which is in calldata as well, whereas the "array view" type would instead point to the first element of the array and have the length on the stack? Makes sense, especially since ``callldataload`` is cheap and stack space is rare. Although I don't really see a type-safety problem with using the same type for both...

Reply:
Ok, type safety depends on what we expose to the user. Using only one stack slot would also make memory and calldata arrays more similar.

Reply:
True. We might want to decide this before ``0.6.0`` because this might lead to a breaking change in one way or another.

Reply:
Created an issue: https://github.com/ethereum/solidity/issues/6220